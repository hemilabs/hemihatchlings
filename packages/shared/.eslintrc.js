module.exports = {
  env: {
    es2021: true,
    node: true
  },
  extends: 'standard-with-typescript',
  overrides: [
  ],
  parserOptions: {
    ecmaVersion: 'latest',
    sourceType: 'module',
    project: ['tsconfig.json'],
    tsconfigRootDir: __dirname
  },
  plugins: [
    'sonarjs'
  ],
  rules: {
    '@typescript-eslint/triple-slash-reference': 'off',
    '@typescript-eslint/space-before-function-paren': 'off',
    '@typescript-eslint/strict-boolean-expressions': 'off',
    '@typescript-eslint/no-extraneous-class': 'off',
    'semi': ["error", "never"],
    'quotes': ["error", "single"],
    'import/extensions': 'off',
    'import/prefer-default-export': 'off',
    indent: ['error', 2],
    'class-methods-use-this': 'off',
    complexity: ['error', 4],
    'sonarjs/cognitive-complexity': ['error', 4],
    'max-depth': ['error', 3],
    'max-statements': ['error', 10],
    'max-lines': ['error', {
      max: 130,
      skipBlankLines: true,
      skipComments: true
    }],
    'max-lines-per-function': ['error', {
      max: 35,
      skipBlankLines: true,
      skipComments: true
    }],
    'max-nested-callbacks': ['error', 3],
    'max-params': ['error', 3],
    'no-useless-constructor': 'off',
    'max-classes-per-file': 'off',
    'lines-between-class-members': [
      'error',
      'always',
      {
        exceptAfterSingleLine: true
      }
    ],
    'no-shadow': 'off',
    '@typescript-eslint/no-shadow': ['error'],
    'max-len': ['error', {
      code: 80
    }],
    overrides: [
      {
        files: ['*.spec.ts'],
        rules: {
          'max-nested-callbacks': ['error', 5],
        }
      }
    ]
  }
}
