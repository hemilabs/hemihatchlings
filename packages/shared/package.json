{
  "name": "@hemihatchlings/shared",
  "version": "0.1.0",
  "description": "Shared package for hemihatchlings Application",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "type": "module",
  "private": true,
  "exports": {
    ".": "./src/index.ts"
  },
  "scripts": {
    "build": "tsc -p ./tsconfig.build.json",
    "lint": "eslint",
    "lint:fix": "eslint --fix",
    "test": "vitest run",
    "test:cov": "vitest run --coverage",
    "test:watch": "vitest watch"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "typescript": "^4.8.2",
    "uuid": "^9.0.0"
  },
  "devDependencies": {
    "@babel/preset-env": "^7.19.0",
    "@types/node": "^18.7.15",
    "@types/uuid": "^8.3.4",
    "@typescript-eslint/eslint-plugin": "^5.0.0",
    "@vitest/coverage-c8": "^0.23.1",
    "eslint": "^8.0.1",
    "eslint-config-standard-with-typescript": "^22.0.0",
    "eslint-plugin-import": "^2.25.2",
    "eslint-plugin-n": "^15.0.0",
    "eslint-plugin-promise": "^6.0.0",
    "eslint-plugin-sonarjs": "^0.15.0",
    "ts-node": "^10.9.1",
    "vite": "^5.0.11",
    "vitest": "^0.23.1"
  }
}
