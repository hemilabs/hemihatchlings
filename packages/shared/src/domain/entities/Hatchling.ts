import { Address } from '../valueObjects/Address'
import { TransactionHash } from '../valueObjects/TransactionHash'
import { Stage } from '../valueObjects/Stage'
import { TransactionStatus } from '../valueObjects/TransactionStatus'
import { TransactionStatusEnum } from '../enums/TransactionStatusEnum'
import { StageEnum } from '../enums/StageEnum'
import { Entity } from '../base/Entity'
import { Uuid } from '../valueObjects/Uuid'
import { Element } from '../valueObjects/Element'

interface HatchlingProps {
  account: Address
  transactionHash: TransactionHash
  transactionStatus: TransactionStatus
  stage: Stage
  element?: Element
}

interface CreateWithStatusProps {
  account: Address
  transactionHash: TransactionHash
  transactionStatus: TransactionStatus
  element?: Element
}

interface CreateWithStageProps {
  account: Address
  transactionHash: TransactionHash
  stage: Stage
  element?: Element
}

export class Hatchling extends Entity<HatchlingProps> {
  private constructor(props: HatchlingProps, id: Uuid) {
    super(props, id)
  }

  static createWithStatus(props: CreateWithStatusProps, id?: Uuid): Hatchling {
    const hatchlingId = id ?? Uuid.create()
    const stage = this.getStageFromStatus(props.transactionStatus)

    return new Hatchling({ ...props, stage }, hatchlingId)
  }

  static createWithStage(props: CreateWithStageProps, id?: Uuid): Hatchling {
    const hatchlingId = id ?? Uuid.create()
    const transactionStatus = this.getStatusFromStage(props.stage)

    return new Hatchling({ ...props, transactionStatus }, hatchlingId)
  }

  private static getStageFromStatus(status: TransactionStatus): Stage {
    if (status.value === TransactionStatusEnum.EthFinality) {
      return Stage.create(StageEnum.Adolescent)
    }

    if (status.value === TransactionStatusEnum.BtcFinality) {
      return Stage.create(StageEnum.Adult)
    }

    return Stage.create(StageEnum.Baby)
  }

  private static getStatusFromStage(stage: Stage): TransactionStatus {
    if (stage.value === StageEnum.Adolescent) {
      return TransactionStatus.create(TransactionStatusEnum.EthFinality)
    }

    if (stage.value === StageEnum.Adult ) {
      return TransactionStatus.create(TransactionStatusEnum.BtcFinality)
    }

    return TransactionStatus.create(TransactionStatusEnum.Minted)
  }

  get account(): Address {
    return this.props.account
  }

  get transactionHash(): TransactionHash {
    return this.props.transactionHash
  }

  get transactionStatus(): TransactionStatus {
    return this.props.transactionStatus
  }

  get stage(): Stage {
    return this.props.stage
  }

  get element(): Element | undefined {
    return this.props.element
  }
}
