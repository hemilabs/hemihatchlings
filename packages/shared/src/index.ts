// Base Code
export * from "./domain/base/DomainError"
export * from "./domain/base/Entity"
export * from "./domain/base/ValueObject"

// Entities
export * from "./domain/entities/Hatchling"

// Enums
export * from "./domain/enums/ElementEnum"
export * from "./domain/enums/StageEnum"
export * from "./domain/enums/TransactionStatusEnum"

// Factories
export * from "./domain/factories/HatchlingFactory"

// Value Objects
export * from "./domain/valueObjects/Address"
export * from "./domain/valueObjects/Element"
export * from "./domain/valueObjects/Stage"
export * from "./domain/valueObjects/TransactionHash"
export * from "./domain/valueObjects/TransactionStatus"
export * from "./domain/valueObjects/Uuid"
