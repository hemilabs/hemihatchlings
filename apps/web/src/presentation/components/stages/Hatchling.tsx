import fireBaby from '../../../assets/fire-baby.png'
import waterBaby from '../../../assets/water-baby.png'
import grassBaby from '../../../assets/grass-baby.png'
import airBaby from '../../../assets/air-baby.png'

import fireAdult from '../../../assets/fire-adult.png'
import waterAdult from '../../../assets/water-adult.png'
import grassAdult from '../../../assets/grass-adult.png'
import airAdult from '../../../assets/air-adult.png'

import lockIcon from '../../../assets/lock-icon.png'
import { ElementEnum, StageEnum } from '@hemihatchlings/shared'

interface HatchlingProps {
  element: ElementEnum,
  stage: StageEnum
}

function Hatchling({ element, stage }:HatchlingProps): JSX.Element {
  const isBaby = (): Boolean => {
    return stage === StageEnum.Baby
  }

  const isAdult = (): Boolean => {
    return stage === StageEnum.Adult
  }

  return (
    <div className="flex flex-row justify-center">

      <div className='flex flex-col'>
        <div className={`${isBaby() ? 'blur' : ''}
          border w-96 items-center`}>
          <img src={fireBaby} alt="Fire Baby"
            className={`${element === ElementEnum.Fire ? '' : 'hidden'}`} />
          <img src={waterBaby} alt="Water Baby"
            className={`${element === ElementEnum.Water ? '' : 'hidden'}`} />
          <img src={grassBaby} alt="Grass Baby"
            className={`${element === ElementEnum.Grass ? '' : 'hidden'}`} />
          <img src={airBaby} alt="Air Baby"
            className={`${element === ElementEnum.Air ? '' : 'hidden'}`} />
        </div>
        <div className="font-monaco text-sm text-white mt-2 text-center
          sm:text-base md:text-lg md:mt-4">
          { isBaby() ? 'ðŸ”„' : 'âœ…' } HATCHING STAGE 1 
        </div>
      </div>
      <div className='flex flex-col'>
        <div className={`${!isAdult() ? 'blur' : ''}
          border w-96 items-center`}>
          <img src={fireAdult} alt="Fire Adult"
            className={`${element === ElementEnum.Fire ? '' : 'hidden'}`} />
          <img src={waterAdult} alt="Water Adult"
            className={`${element === ElementEnum.Water ? '' : 'hidden'}`} />
          <img src={grassAdult} alt="Grass Adult"
            className={`${element === ElementEnum.Grass ? '' : 'hidden'}`} />
          <img src={airAdult} alt="Air Adult"
            className={`${element === ElementEnum.Air ? '' : 'hidden'}`} />
        </div>

        <div className="font-monaco text-sm text-white mt-2 text-center
        sm:text-base md:text-lg md:mt-4">
          { !isAdult() ? 'ðŸ”„' : 'âœ…' } HATCHING STAGE 2 
        </div>
      </div>

      <img src={lockIcon}
          alt="Lock Icon"
          className='absolute m-auto top-0 bottom-0 left-0 right-0' />

    </div>
    
  )
}

export default Hatchling
