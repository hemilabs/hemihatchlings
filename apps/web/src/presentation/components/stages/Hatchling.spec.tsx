import { describe, it, expect} from 'vitest'
import { RenderResult, render } from '@testing-library/react'
import Hatchling from './Hatchling'
import { ElementEnum, StageEnum } from '@hemihatchlings/shared'

interface HatchlingProps {
  element: ElementEnum,
  stage: StageEnum
}

const renderComponent = ({ element, stage }:HatchlingProps): RenderResult => {
    return render(
      <Hatchling element={element} stage={stage} />
    )
  }

describe('src/presentation/Hatchling', () => {
  it('renders correctly', () => {
    const { asFragment } = renderComponent({
      element: ElementEnum.Fire,
      stage: StageEnum.Egg
    })

    expect(asFragment()).toMatchSnapshot()
  })
})
