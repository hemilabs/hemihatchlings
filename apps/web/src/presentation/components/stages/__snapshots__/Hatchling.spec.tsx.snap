// Vitest Snapshot v1

exports[`src/presentation/Hatchling > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row justify-center"
  >
    <div
      class="flex flex-col"
    >
      <div
        class="blur
          border w-96 items-center"
      >
        <img
          alt="Fire Baby"
          class=""
          src="/src/assets/fire-baby.png"
        />
        <img
          alt="Water Baby"
          class="hidden"
          src="/src/assets/water-baby.png"
        />
        <img
          alt="Grass Baby"
          class="hidden"
          src="/src/assets/grass-baby.png"
        />
        <img
          alt="Air Baby"
          class="hidden"
          src="/src/assets/air-baby.png"
        />
      </div>
      <div
        class="font-monaco text-sm text-white mt-2 text-center sm:text-base md:text-lg md:mt-4"
      >
        ðŸ”„ HATCHING STAGE 1
      </div>
    </div>
    <div
      class="flex flex-col"
    >
      <div
        class="blur
          border w-96 items-center"
      >
        <img
          alt="Fire Adult"
          class=""
          src="/src/assets/fire-adult.png"
        />
        <img
          alt="Water Adult"
          class="hidden"
          src="/src/assets/water-adult.png"
        />
        <img
          alt="Grass Adult"
          class="hidden"
          src="/src/assets/grass-adult.png"
        />
        <img
          alt="Air Adult"
          class="hidden"
          src="/src/assets/air-adult.png"
        />
      </div>
      <div
        class="font-monaco text-sm text-white mt-2 text-center sm:text-base md:text-lg md:mt-4"
      >
        ðŸ”„ HATCHING STAGE 2
      </div>
    </div>
    <img
      alt="Lock Icon"
      class="absolute m-auto top-0 bottom-0 left-0 right-0"
      src="/src/assets/lock-icon.png"
    />
  </div>
</DocumentFragment>
`;
